### DISCUSSION TITLE
Using replace in SQL functions in .drift files

### DISCUSSION THREAD

**Discussion by frankvollebregt** [simolus3/drift#2277] [Q&A] [answered]
## My specific context
I'm writing a query that retrieves information about a person from the `persons` table in my database. This includes their phone number from a joined `phonenumbers` table. Since a person may have multiple phone numbers (1-n), but I don't want to return the whole person multiple times, I was trying to concatenate the different phone numbers into one, also removing duplicates. I came across a forum post on the SQLite forums:
https://sqlite.org/forum/forumpost/4d51fbce1abf594c

Which suggests a one-line way to achieve this by using rtrim, replace and group_concat. When I try to build using this, Drift warns me that `replace` is a reserved keyword, but doesn't seem to understand that I'm trying to call the `replace` function. 

## The actual issue
The `replace` operator seems to be unsupported (but understood as being a reserved keyword). A simple case where this would also occur could be as follows:
```sql
CREATE TABLE books (
  id INT NOT NULL PRIMARY KEY,
  title TEXT NULL
);

-- Get the titles as they are
getTitles: SELECT title FROM books;

-- Replace spaces with underscores
getTitlesSnakeCase: SELECT replace(title, ' ', '_') FROM books;
```
After which the `build_runner` points to the word `replace` and remarks:
```
Could not parse this expression. Note: This is a reserved keyword, you can escape it in double ticks
```
Is this not (yet) supported? If not, is there a specific reason for that or did it simply not have the priority?

## Note on using this in my context
For my specific use case, I've since realized that a nicer way to do it is using Drift [LIST subqueries](https://sqlite.org/forum/forumpost/4d51fbce1abf594c), which brought about a different issue altogether, but I'll create a separate issue for this.

**Comment by North101** [comment, marked as answer]
does this work?
```sql
getTitlesSnakeCase: SELECT replace(title, ' ', '_') AS title FROM books;
```

**Reply by frankvollebregt** [reply]
Thanks for taking a look. It does not, unfortunately.
```
[WARNING] drift_dev:drift_dev on lib/data/tables.drift:
line 40, column 28: Could not parse this expression. Note: This is a reserved keyword, you can escape it in double ticks
   ╷
40 │ getTitlesSnakeCase: SELECT replace(title, ' ', '_') AS title FROM books;
   │                            ^^^^^^^
   ╵
```
I guess the thought was that the exported Dart name could be causing the reserved keyword warning, rather than the expression itself?

**Reply by simolus3** [reply]
> `Note: This is a reserved keyword, you can escape it in double ticks`

This actually works for functions as well, both sqlite3 and `drift_dev` will happily accept `SELECT "replace"(title, ' ', '_') FROM books`.

I'm not sure if this should be fixed - of course, we should stay as close to sqlite3 as possible for compatibility. After all, that is the primary point of queries analyzed at compile time. But sqlite3 is very quirky in the way it interprets keywords as identifiers in some cases - this behavior is mostly undocumented and really hard to reproduce reliably. So for convenience and safety, I think I'll call the current behavior working as intended.

**Reply by frankvollebregt** [reply]
Oh, maybe it'd be better to change the reported warning a bit then. I interpreted it as "if you want to use replace as a name for some column, you can escape it so you won't accidentally use the function", but in Drift I will actually need to escape it to use it as a function

**Reply by frankvollebregt** [reply]
Hey @simolus3,

The issue itself is resolved, but I think the `Note` provided is not very clear (at least it wasn't for me): It implied to me that I could use ticks to use `replace` in my variable names, not that I had to use them in order to use the `replace` function.

Additionally, when I read double ticks I think about this tick ` whereas the one that is intended " is what I would normally call a quote.

I'm not entirely sure for which keywords this message is used. If it's only for these [built-in functions](https://sqlite.org/lang_corefunc.html), I propose changing the note to something like the following:
```
Note: this is a keyword, you can use this function in your query by wrapping it in double quotes.
```
If it's worth going through the effort at all, considering this is not something most users will ever use/run into...

**Reply by simolus3** [reply]
I agree, I've changed the error message to give better advice when the keyword is followed by a left parenthesis in e2136236d9b56e78a88414700854bf1fb2141da8.
